File: src/Cpp/23_Wyrazenia_regularne/Zad11.cpp
#include <cassert>
#include <regex>
#include <string>

//Otrzymujesz napis reprezentujacy sciezke pliku. Odfiltruj nazwe pliku ze sciezki.

std::string nazwaPliku(std::string& sciezka)
{
    const std::regex wzorzec("(.*[\\\\/]+)?([^\\.]*)(.*)?");
    std::smatch wynik;

    if (std::regex_match(sciezka, wynik, wzorzec) && wynik.size() > 1)
        return wynik[2].str();

    return "";
}

void test1()
{
    std::string sciezka = "C:\\Windows\\explorer.exe";
    std::string wynik = "explorer";

    assert(nazwaPliku(sciezka) == wynik);
}

void test2()
{
    std::string sciezka = "/home/user/var/log/syslog";
    std::string wynik = "syslog";
    assert(nazwaPliku(sciezka) == wynik);
}

void test3()
{
    std::string sciezka = "test.jpg.png.mp3.mp5.html.txt";
    std::string wynik = "test";
    assert(nazwaPliku(sciezka) == wynik);
}

void test4()
{
    std::string sciezka = "test";
    std::string wynik = "test";
    assert(nazwaPliku(sciezka) == wynik);
}

int main()
{

    test1();
    test2();
    test3();
    test4();

    return 0;
}

File: src/python/23_wyrazenia_regularne/zad11.py
"""
ZAD-11 — Nazwa pliku bez rozszerzenia

**Poziom:** ★★☆
**Tagi:** `regex`, `string`, `ścieżki`

### Treść

Otrzymujesz napis reprezentujący pełną ścieżkę do pliku. Wyodrębnij nazwę pliku (bez katalogów) i usuń rozszerzenie (część po ostatniej kropce). Zwróć samą nazwę bez rozszerzenia.

Ścieżka może zawierać separator `\` lub `/`.

### Wejście

Jedna linia:

* `sciezka`

### Wyjście

Jedna linia:

* `nazwa_pliku_bez_rozszerzenia`

### Przykład

**Wejście:**

```
C:\my-long\path_directory\file.html
```

**Wyjście:**

```
file
```

"""

import re


def nazwa_pliku(sciezka):
    """
    Funkcja zwraca nazwe pliku z podanej sciezki.
    """
    nazwa = re.search(r'[^\\/:*?"<>|]+$', sciezka).group(0)

    # usun rozszerzenie pliku
    while re.sub(r"\.[^.]+$", "", nazwa) is not nazwa:
        nazwa = re.sub(r"\.[^.]+$", "", nazwa)
    return nazwa


def test_nazwa_pliku():
    assert nazwa_pliku("C:\\Windows\\explorer.exe") == "explorer"
    assert nazwa_pliku("/home/user/var/log/syslog") == "syslog"
    assert nazwa_pliku("test.jpg.png.mp3.mp5.html.txt") == "test"
    assert nazwa_pliku("test") == "test"


if __name__ == "__main__":
    test_nazwa_pliku()
